Runtime/Scripts/PersistentConnection.cs

```diff
@@ -131,192 +217,193 @@ IEnumerator OnIntervalEdgeEventsLocation(EdgeEventsConnection connection)
           }
         }
+        else
+        {
+          PropagateError(FindCloudletEventTrigger.LatencyTooHigh, "FindCloudletPerformanceMode failed");
+        }
+      }
+    }
 
-    private void OnApplicationFocus(bool focus)
+    async void CompareLatencies(Statistics receivedStats)
     {
-      if (locationUpdatesRunning)
+      AppPort appPort;
+      Site site;
+      var netTest = new NetTest(integration.matchingEngine);
+      if (hasTCPPorts)
+      {
+        appPort = integration.GetAppPort(LProto.Tcp, config.latencyTestPort == 0 ? 0 : config.latencyTestPort);
+        site = new Site { host = appPort.FqdnPrefix + integration.latestFindCloudletReply.Fqdn, port = appPort.PublicPort, testType = TestType.CONNECT };
+      }
+      else
       {
-        StartCoroutine(OnIntervalEdgeEventsLocation(integration.matchingEngine.EdgeEventsConnection));
+        appPort = integration.GetAppPort(LProto.Udp);
+        site = new Site { host = appPort.FqdnPrefix + integration.latestFindCloudletReply.Fqdn, port = appPort.PublicPort, testType = TestType.PING };
       }
-      if (latencyUpdatesRunning)
+      await netTest.TestSite(site);
+      double normalizedLatency = receivedStats.Avg - (receivedStats.Avg * config.performanceSwitchMargin);
+      if (site.average < normalizedLatency)
       {
-        StartCoroutine(OnIntervalEdgeEventsLatency(integration.matchingEngine.EdgeEventsConnection, integration.GetHost()));
+        PropagateSuccess(FindCloudletEventTrigger.LatencyTooHigh, integration.latestFindCloudletReply);
+      }
+      else
+      {
+        PropagateError(FindCloudletEventTrigger.LatencyTooHigh,
+          "Cloudlet obtained from FindCloudletPerformanceMode have higher latency than the previous cloudlet");
       }
     }
 
-    private void OnApplicationPause(bool pause)
+    void CleanUp()
     {
-      StopAllCoroutines();// stop edge events streaming
+      latencyUpdatesCounter = 0;
+      latencyUpdatesRunning = false;
+      locationUpdatesCounter = 0;
+      locationUpdatesRunning = false;
+      location = null;
     }
 
-    private void OnApplicationQuit()
+    async void PropagateSuccess(FindCloudletEventTrigger trigger, FindCloudletReply newCloudlet)
     {
-      StopAllCoroutines();
+      Logger.Log("Received NewCloudlet from the server triggered by" + trigger);
+      FindCloudletEvent findCloudletEvent = new FindCloudletEvent();
+      findCloudletEvent.trigger = trigger;
+      findCloudletEvent.newCloudlet = newCloudlet;
+      EdgeEventsStatus eventStatus = new EdgeEventsStatus(Status.success);
+      try
+      {
+        integration.NewFindCloudletHandler(eventStatus, findCloudletEvent);
+      }
+      catch (NullReferenceException)
+      {
+        Debug.LogError("Missing NewCloudletHandler, " +
+          "Listen to EdgeEvents by using \"mobiledgexIntegration.NewCloudletHandler += HandleNewFindCloudlet;\"");
```

Comment:
You should probably validate that NewCloudletHandler exists in your validate function as well

Reply:
I researched this and probably I can't validate that before triggering an event. 
@lgarner  Do you know of a way that I can check of the Action Listeners Count or Existance **without bothering the developer with an event**-something that is not a performance killer **no reflection** -?

Reply:
Will do this in another PR if I found a solution.

Reply:
please file a jira ticket to track this

Reply:
Done https://mobiledgex.atlassian.net/browse/EDGECLOUD-5172